python/ray/dag/compiled_dag_node.py

```diff
@@ -35,84 +37,89 @@ def do_allocate_channel(self, buffer_size_bytes: int, num_readers: int = 1) -> C
+                    task.input_channels.append(inp)
+                    task.input_channel_idxs.append(idx)
+                    task.resolved_inputs.append(None)
+                else:
+                    task.resolved_inputs.append(inp)
 
-        self._input_reader: ReaderInterface = SynchronousReader(input_channels)
-        self._output_writer: WriterInterface = SynchronousWriter(self._output_channel)
-        self._input_reader.start()
-        self._output_writer.start()
+            input_reader: ReaderInterface = SynchronousReader(task.input_channels)
+            output_writer: WriterInterface = SynchronousWriter(
+                task.output_channel
+            )
+            self._method_to_input_reader[task.method_name] = input_reader
+            self._method_to_output_writer[task.method_name] = output_writer
+
+            input_reader.start()
+            output_writer.start()
 
         while True:
-            res = self._input_reader.begin_read()
-
-            for idx, output in zip(input_channel_idxs, res):
-                resolved_inputs[idx] = output
-
-            try:
-                output_val = method(*resolved_inputs)
-            except Exception as exc:
-                backtrace = ray._private.utils.format_error_message(
-                    "".join(
-                        traceback.format_exception(type(exc), exc, exc.__traceback__)
-                    ),
-                    task_exception=True,
-                )
-                wrapped = RayTaskError(
-                    function_name="do_exec_compiled_task",
-                    traceback_str=backtrace,
-                    cause=exc,
-                )
-                self._output_writer.write(wrapped)
-            else:
-                self._output_writer.write(output_val)
-            finally:
-                self._input_reader.end_read()
+            for task in tasks:
+                input_reader = self._method_to_input_reader[task.method_name]
+                output_writer = self._method_to_output_writer[task.method_name]
+                res = input_reader.begin_read()
+
+                for idx, output in zip(task.input_channel_idxs, res):
+                    task.resolved_inputs[idx] = output
+
+                try:
+                    output_val = method(*task.resolved_inputs)
+                except Exception as exc:
+                    backtrace = ray._private.utils.format_error_message(
+                        "".join(
+                            traceback.format_exception(
+                                type(exc), exc, exc.__traceback__
+                            )
+                        ),
+                        task_exception=True,
+                    )
+                    wrapped = RayTaskError(
+                        function_name="do_exec_tasks",
+                        traceback_str=backtrace,
+                        cause=exc,
+                    )
+                    output_writer.write(wrapped)
```

Comment:
For cases where the output is passed as an input to a task on the same actor, it'd be good to have some kind of LocalChannel to avoid needing to serialize the data and synchronize. But you can also leave this as a TODO!

Reply:
Added a TODO.